setup.py

```diff
@@ -51,13 +51,15 @@
 
     abs_path = os.path.dirname(os.path.abspath(__file__)) + '/'
 
-    def call(args):
-        subprocess.check_call(args, cwd=abs_path, env=configure_env)
+    if not os.path.isfile(abs_path + 'src/.libs/libwallycore' + ('.so' if build_shared else '.a')):
```

Comment:
`.so` isn't the suffix of shared libraries on macOS.

@jgriffiths: Do you have any suggestion of how to check here whether the shared library has already been built?

In my Gentoo ebuild, I've been having to use `sed` to delete this entire `if not is_windows:` block because we build the Python module (conditionally, depending on the `python` `USE` flag) after the C library has already been built, but `setup.py` always tries to rebuild it. Worse, Gentoo's Python build machinery makes multiple calls to `setup.py` since building and installing are separate ebuild phases, yet this `setup.py` does the build unconditionally every time it's executed. So that's my motivation for making the build steps here conditional depending on whether the needed library type already exists.

Reply:
It should be possible to determine what args setup.py was called with I think, can you give me the calls that are made (and the arguments) by the Gentoo machinery?

Reply:
> can you give me the calls that are made (and the arguments) by the Gentoo machinery?

Hmm, actually no, I can't. It's complicated by the fact that `setup.py` is considered deprecated, so Gentoo now uses [gpep517](https://pypi.org/project/gpep517/) in legacy setuptools mode to drive the execution of the code in `setup.py` for projects that have not yet graduated to a PEP517-compliant build system. The extra layer of abstraction makes it all very unclear to me. (I'm most definitely not a Python guy.)

Reply:
If you don't like the `isfile` conditional here, I'm fine to drop it, and my ebuild will continue to sed out the code block containing these build commands. Note, however, that that would make it awkward for anyone else who is trying to achieve the "full dynamic link" outcome, as the unmodified `setup.py` will undo the effect of `--with-system-secp256k1` and rebuild the library with embedded secp256k1 code.

Reply:
I'm focused on getting the other PR in first, but in general I'm less worried about building the python wheel being more complicated in this case (not to say I'm not willing to add support).

Reply:
In general, what is the logic behind having the Python module be responsible for building the C library? That's an unexpected inversion of responsibility relative to historical norms, wherein the native library is built standalone, and then wrappers for other languages are built to link with the native library.

Reply:
We've had many problems in the past not using the Python build machinery, as Python doesn't care about historic norms. Python expects the c code for c-extensions to be built by setup.py, and although you can compile just the wrapping sources and link with a DSO as you will on Gentoo, that only works for distributions and can't work for PyPI installs as (a) pip installing wallycore would import an unusable module unless you had a compatible wally shared lib installed and (b) it circumvents the security of using a requirements.txt version hash because you can't be sure which libwallycore.so is going to be loaded.

Reply:
> pip installing wallycore would import an unusable module unless you had a compatible wally shared lib installed

That's exactly what I would expect, to be frank. Trying to use a wrapper library without also having installed the underlying library that it wraps should be expected to fail.

> it circumvents the security of using a requirements.txt version hash because you can't be sure which libwallycore.so is going to be loaded.

This amuses me because I see it just the opposite: pinning a version of libwally-core in the Python wrapper means that the Python wrapper *cannot* benefit from security fixes in the underlying libwally-core unless the Python wrapper also gets updated. A user may upgrade libwally-core to resolve a security issue and expect that all code which uses libwally-core will be fixed. That's certainly the assumption on Gentoo, which is why static linking is so insidious and harmful to security. I guess it's like tabs versus spaces: we'll never agree. (Incidentally, your code base is horrible — I keep having to hit SPACE SPACE SPACE SPACE in my editors since you don't use sane tabulation. 😁)

So what do you want to do about the matter at hand? Like I said, I'm okay to rip out the conditional and keep patching it downstream, which sounds like the right thing to do since linking a Python wrapper module dynamically to the library that it wraps sounds like practice unsanctioned by the Python gods.